tools/PlateManager/PlateFile2WorkItemGenerator.cs

```diff
@@ -0,0 +1,85 @@
+﻿using Microsoft.Extensions.Logging;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WWT.Azure;
+using WWTWebservices;
+
+namespace PlateManager
+{
+    internal sealed class PlateFile2WorkItemGenerator : PlateFileWorkItemGeneratorBase, IWorkItemGenerator
+    {
+        private readonly AzurePlateTilePyramid _pyramid;
+        private readonly ILogger<PlateFile2WorkItemGenerator> _logger;
+        
+        public PlateFile2WorkItemGenerator(AzurePlateTilePyramid pyramid, ILogger<PlateFile2WorkItemGenerator> logger)
+        {
+            _pyramid = pyramid;
+            _logger = logger;
+        }
+
+        public IEnumerable<Func<int, int, Task>> GenerateWorkItems(string plateFile, string baseUrl, CancellationToken token)
+        {
+            var filepart = Path.GetFileNameWithoutExtension(plateFile);
+            var azureContainer = Path.GetFileName(plateFile).ToLowerInvariant();
+            
+            // Handle thumbnails if one exists alongside the plate file
+            string thumbnail = GetThumbnailName(plateFile);
+            if (File.Exists(thumbnail))
+            {
+                _logger.LogTrace("Adding task for thumbnail {Path}", thumbnail);
+                Task UploadThumbnail(int count, int total) => _pyramid.SaveStreamAsync(GetFileStream(thumbnail), azureContainer, GetThumbnailBlobName(filepart), token);
+                yield return UploadThumbnail;
+            }
+
+            // Re-write entries in the WTML file to point to the new location for the images
+            // to point to Azure if the WTML file exists alongside the plate file
+            string wtmlfile = GetWtmlName(plateFile);
+            if (File.Exists(wtmlfile))
+            {
+                string wtmlFileOut = wtmlfile.Replace(".wtml", ".azure.wtml");
+                string wtmldata = UpdateWtmlEntries(File.ReadAllText(wtmlfile), filepart, baseUrl, azureContainer);
+                File.WriteAllText(wtmlFileOut, wtmldata);
+            }
+
+            var plateFile2 = new PlateFile2(plateFile);
+            var directoryEntries = plateFile2.GetEntries().ToList();
```

Comment:
Not sure how many there are, but do we need to precalculate this or just lazily get them in the foreach?

Reply:
I went ahead and evaluated the IEnumerable here since I wanted a maximum count of items anyway for the logging message and would be traversing the nodes.